Description: when suspending or hibernating, use the new ubuntu-lock-on-suspend preference.
Author: Marc Deslauriers <marc.deslauriers@canonical.com>
Forwarded: no, likely a Ubuntu-specific preference
Bug-Ubuntu: https://bugs.launchpad.net/bugs/938076
Bug-Ubuntu: https://bugs.launchpad.net/bugs/847814

Index: cinnamon-settings-daemon-3.6.1/plugins/power/csd-power-manager.c
===================================================================
--- cinnamon-settings-daemon-3.6.1.orig/plugins/power/csd-power-manager.c	2012-10-05 13:56:44.000000000 +0200
+++ cinnamon-settings-daemon-3.6.1/plugins/power/csd-power-manager.c	2012-10-26 10:20:11.739005253 +0200
@@ -2249,9 +2249,6 @@
         GError *error = NULL;
         CsdPowerActionType action_type;
 
-        /* maybe lock the screen if the lid is closed */
-        lock_screensaver (manager);
-
         /* we have different settings depending on AC state */
         if (up_client_get_on_battery (manager->priv->up_client)) {
                 action_type = g_settings_get_enum (manager->priv->settings,
@@ -2268,6 +2265,9 @@
                         g_warning ("to prevent damage, now forcing suspend");
                         do_power_action_type (manager, CSD_POWER_ACTION_SUSPEND);
                         return;
+                } else {
+                        /* maybe lock the screen if the lid is closed */
+                        lock_screensaver (manager);
                 }
         }
 
@@ -3499,7 +3499,31 @@
                         UpSleepKind sleep_kind,
                         CsdPowerManager *manager)
 {
-        lock_screensaver (manager);
+        gboolean do_lock;
+
+        do_lock = g_settings_get_boolean (manager->priv->settings_screensaver,
+                                          "ubuntu-lock-on-suspend");
+        if (!do_lock)
+                return;
+
+        if (manager->priv->screensaver_proxy != NULL) {
+                g_debug ("doing cinnamon-screensaver lock");
+                g_dbus_proxy_call (manager->priv->screensaver_proxy,
+                                   "Lock",
+                                   NULL, G_DBUS_CALL_FLAGS_NONE, -1,
+                                   NULL, NULL, NULL);
+        } else {
+                /* connect to the screensaver first */
+                g_dbus_proxy_new_for_bus (G_BUS_TYPE_SESSION,
+                                          G_DBUS_PROXY_FLAGS_DO_NOT_LOAD_PROPERTIES,
+                                          NULL,
+                                          GS_DBUS_NAME,
+                                          GS_DBUS_PATH,
+                                          GS_DBUS_INTERFACE,
+                                          NULL,
+                                          sleep_cb_screensaver_proxy_ready_cb,
+                                          manager);
+        }
 }
 
 static void
